creator-node/src/fileManager.js

```diff
@@ -192,52 +223,69 @@ async function saveFileForMultihashToFS (req, multihash, expectedStoragePath, ga
             continue
           }
         }
 
         if (!response || !response.data) {
+          decisionTree.push({ stage: `Couldn't find files on other creator nodes, after trying URLs`, vals: null, time: Date.now() })
           throw new Error(`Couldn't find files on other creator nodes, after trying URLs: ${gatewayUrlsMapped.toString()}`)
         }
 
         // Write file to disk
         await Utils.writeStreamToFileSystem(response.data, expectedStoragePath)
         fileFound = true
 
+        decisionTree.push({ stage: 'Wrote file to file system after fetching from gateway', vals: expectedStoragePath, time: Date.now() })
         req.logger.info(`wrote file to ${expectedStoragePath}`)
       } catch (e) {
+        decisionTree.push({ stage: `Failed to retrieve file for multihash from other creator node gateways`, vals: e.message, time: Date.now() })
         throw new Error(`Failed to retrieve file for multihash ${multihash} from other creator node gateways: ${e.message}`)
       }
     }
 
     // file was not found on ipfs or any gateway
     if (!fileFound) {
+      decisionTree.push({ stage: 'Failed to retrieve file for multihash after trying ipfs & other creator node gateways', vals: multihash, time: Date.now() })
       throw new Error(`Failed to retrieve file for multihash ${multihash} after trying ipfs & other creator node gateways`)
     }
 
     // verify that the contents of the file match the file's cid
     try {
+      decisionTree.push({ stage: 'About to verify the file contents for the CID', vals: multihash, time: Date.now() })
       const ipfs = req.app.get('ipfsLatestAPI')
       const content = fs.createReadStream(expectedStoragePath)
       for await (const result of ipfs.add(content, { onlyHash: true, timeout: 10000 })) {
         if (multihash !== result.cid.toString()) {
+          decisionTree.push({ stage: `File contents don't match IPFS hash multihash`, vals: result.cid.toString(), time: Date.now() })
           throw new Error(`File contents don't match IPFS hash multihash: ${multihash} result: ${result.cid.toString()}`)
         }
       }
+      decisionTree.push({ stage: 'Successfully verified the file contents for the CID', vals: multihash, time: Date.now() })
     } catch (e) {
+      decisionTree.push({ stage: `Error during content verification for multihash`, vals: multihash, time: Date.now() })
       throw new Error(`Error during content verification for multihash ${multihash} ${e.message}`)
     }
 
+    _printDecisionTreeObj(req, decisionTree)
     return expectedStoragePath
   } catch (e) {
+    decisionTree.push({ stage: `saveFileForMultihashToFS error`, vals: e.message, time: Date.now() })
+    _printDecisionTreeObj(req, decisionTree)
     throw new Error(`saveFileForMultihashToFS - ${e}`)
   }
 }
 
+const _printDecisionTreeObj = (req, decisionTree) => {
+  req.logger.info('saveFileForMultihashToFS decision tree', JSON.stringify(decisionTree))
```

Comment:
my one concern with this would be if the log call fails, which could happen if the `JSON.stringify()` op fails due to something like circular JSON etc if one of the fields is some large nested object

not sure best way to handle, could just try-catch here but 🤷‍♂️

Reply:
you right, added

Reply:
perf